broker-core/src/test/java/io/zeebe/broker/subscription/message/state/MessageSubscriptionStateTest.java

```diff
@@ -0,0 +1,325 @@
+    state.visitSubscriptions(
+        wrapString("messageName"),
+        wrapString("correlationKey"),
+        s -> {
+          keys.add(s.getElementInstanceKey());
+          return false;
+        });
+
+    // then
+    assertThat(keys).hasSize(2).contains(1L, 2L);
+  }
+
+  @Test
+  public void shouldNoVisitMessageSubscriptionBeforeTime() {
+    // given
+    final MessageSubscription subscription1 = subscriptionWithElementInstanceKey(1L);
+    state.put(subscription1);
+    state.updateSentTime(subscription1, 1_000);
+
+    final MessageSubscription subscription2 = subscriptionWithElementInstanceKey(2L);
+    state.put(subscription2);
+    state.updateSentTime(subscription2, 3_000);
+
+    // then
+    final List<Long> keys = new ArrayList<>();
+    state.visitSubscriptionBefore(1_000, s -> keys.add(s.getElementInstanceKey()));
+
+    assertThat(keys).isEmpty();
+  }
+
+  @Test
+  public void shouldVisitMessageSubscriptionBeforeTime() {
+    // given
+    final MessageSubscription subscription1 = subscriptionWithElementInstanceKey(1L);
+    state.put(subscription1);
+    state.updateSentTime(subscription1, 1_000);
+
+    final MessageSubscription subscription2 = subscriptionWithElementInstanceKey(2L);
+    state.put(subscription2);
+    state.updateSentTime(subscription2, 3_000);
+
+    // then
+    final List<Long> keys = new ArrayList<>();
+    state.visitSubscriptionBefore(2_000, s -> keys.add(s.getElementInstanceKey()));
+
+    assertThat(keys).hasSize(1).contains(1L);
+  }
+
+  @Test
+  public void shouldFindMessageSubscriptionBeforeTimeInOrder() {
+    // given
+    final MessageSubscription subscription1 = subscriptionWithElementInstanceKey(1L);
+    state.put(subscription1);
+    state.updateSentTime(subscription1, 1_000);
+
+    final MessageSubscription subscription2 = subscriptionWithElementInstanceKey(2L);
+    state.put(subscription2);
+    state.updateSentTime(subscription2, 2_000);
+
+    // then
+    final List<Long> keys = new ArrayList<>();
+    state.visitSubscriptionBefore(3_000, s -> keys.add(s.getElementInstanceKey()));
+
+    assertThat(keys).hasSize(2).containsExactly(1L, 2L);
+  }
+
+  @Test
+  public void shouldNotVisitMessageSubscriptionIfSentTimeNotSet() {
+    // given
+    final MessageSubscription subscription1 = subscriptionWithElementInstanceKey(1L);
+    state.put(subscription1);
+    state.updateSentTime(subscription1, 1_000);
+
+    final MessageSubscription subscription2 = subscriptionWithElementInstanceKey(2L);
+    state.put(subscription2);
+
+    // then
+    final List<Long> keys = new ArrayList<>();
+    state.visitSubscriptionBefore(2_000, s -> keys.add(s.getElementInstanceKey()));
+
+    assertThat(keys).hasSize(1).contains(1L);
+  }
+
+  @Test
+  public void shouldUpdateMessageSubscriptionSentTime() {
+    // given
+    final MessageSubscription subscription = subscriptionWithElementInstanceKey(1L);
+    state.put(subscription);
+
+    // when
+    state.updateSentTime(subscription, 1_000);
```

Comment:
Can we have a test on update the sent time twice and check if the there is still one subscription?